pkg/types/spec/utils.go

```diff
@@ -0,0 +1,763 @@
+			model.ModelPath, err = absolutePath(model.ModelPath, projectFiles.ProjectDir())
+			if err != nil {
+				return []CuratedModelResource{}, err
+			}
+		}
+
+		model.ModelPath = s.EnsureSuffix(model.ModelPath, "/")
+
+		models = append(models, CuratedModelResource{
+			ModelResource: &userconfig.ModelResource{
+				Name:         model.Name,
+				ModelPath:    model.ModelPath,
+				SignatureKey: model.SignatureKey,
+			},
+			S3Path: isS3Path,
+		})
+	}
+
+	return models, nil
+}
+
+// Retrieves the model objects found in the S3/local path directory.
+//
+// The model name is determined from the objects' names found in the path directory.
+// Path can either be an S3 path or a local system path - in the latter case, the returned paths will be in absolute form.
+func retrieveModelsResourcesFromPath(path string, projectFiles ProjectFiles, awsClient *aws.Client) ([]userconfig.ModelResource, error) {
+	models := []userconfig.ModelResource{}
+
+	if aws.IsValidS3Path(path) {
+		awsClientForBucket, err := aws.NewFromClientS3Path(path, awsClient)
+		if err != nil {
+			return models, err
+		}
+
+		if isDir, err := awsClientForBucket.IsS3PathDir(path); err != nil {
+			return models, err
+		} else if isDir {
+			modelPaths, _, err := awsClientForBucket.GetNLevelsDeepFromS3Path(path, 1, false, pointer.Int64(20000))
+			if err != nil {
+				return models, err
+			}
+			var bucket string
+			bucket, _, err = aws.SplitS3Path(path)
+			if err != nil {
+				return models, err
+			}
+
+			for _, modelPath := range modelPaths {
+				models = append(models, userconfig.ModelResource{
+					Name:      filepath.Base(modelPath),
+					ModelPath: aws.S3Path(bucket, modelPath),
+				})
+			}
+		} else {
+			return models, ErrorS3DirNotFound(path)
+		}
+
+	} else {
+		var err error
+		path, err = absolutePath(path, projectFiles.ProjectDir())
+		if err != nil {
+			return models, err
+		}
+
+		var fi os.FileInfo
+		fi, err = os.Stat(path)
+		if err != nil {
+			return models, ErrorInvalidPath(path)
+		}
+		if !fi.Mode().IsDir() {
+			return models, ErrorInvalidDirPath(path)
+		}
+
+		var file *os.File
+		file, err = os.Open(path)
+		if err != nil {
+			return models, err
+		}
+
+		var modelObjects []string
+		modelObjects, err = file.Readdirnames(0)
+		if err != nil {
+			return models, err
+		}
+
+		for _, modelName := range modelObjects {
+			models = append(models, userconfig.ModelResource{
+				Name:      modelName,
+				ModelPath: filepath.Join(path, modelName),
+			})
+		}
+	}
+
+	return models, nil
+}
+
+// getPythonVersionsFromS3Paths checks that the path contains a valid S3 directory for versioned Python models:
+// - model-name
+// 		- 1523423423/ (version prefix, usually a timestamp)
+// 			- *
+// 		- 2434389194/ (version prefix, usually a timestamp)
+//			- *
+// 		...
+func getPythonVersionsFromS3Paths(modelPath string, awsClientForBucket *aws.Client) ([]int64, error) {
+	if yes, err := awsClientForBucket.IsS3PathDir(modelPath); !yes && err == nil {
```

Comment:
Good call, yeah. `err == nil` will still be needed in `err == nil && !isDir`, because if `isDir` is set to false, then `err` can be nil or not nil, and depending on this, a different error should be returned in each case.